File: w3ggo/raw.go
package w3ggo

import (
	"bytes"
	"compress/zlib"
	"io"
	"strings"
)

type RawHeader struct {
	compressedSize   uint32
	headerVersion    string
	decompressedSize uint32
	blockCount       uint32
}

type RawSubHeader struct {
	gameIdentifier  string
	version         uint32
	buildNo         uint16
	replayLengthMS  uint32
}

type dataBlock struct {
	blockSize            uint16
	blockDecompressedSize uint16
	blockContent         []byte
}

func parseRaw(input []byte) (*RawHeader, *RawSubHeader, []dataBlock) {
	magic := []byte("Warcraft III recorded game")
	start := findSubsequence(input, magic)
	if start < 0 {
		return nil, nil, nil
	}

	p := newBufParser(input)
	p.setPos(start)
	p.readZeroTermString() // consume the null-terminated magic string
	p.skip(4)              // unknown 4 bytes

	compressedSize, ok := p.readU32LE()
	if !ok {
		return nil, nil, nil
	}
	headerVersion := p.readStringOfLengthAsHex(4)
	decompressedSize, ok := p.readU32LE()
	if !ok {
		return nil, nil, nil
	}
	blockCount, ok := p.readU32LE()
	if !ok {
		return nil, nil, nil
	}

	header := &RawHeader{
		compressedSize:   compressedSize,
		headerVersion:    headerVersion,
		decompressedSize: decompressedSize,
		blockCount:       blockCount,
	}

	// Subheader
	gameIdentifier := p.readStringOfLengthUTF8(4)
	version, ok := p.readU32LE()
	if !ok {
		return nil, nil, nil
	}
	buildNo, ok := p.readU16LE()
	if !ok {
		return nil, nil, nil
	}
	p.skip(2)
	replayLengthMS, ok := p.readU32LE()
	if !ok {
		return nil, nil, nil
	}
	p.skip(4)

	subHeader := &RawSubHeader{
		gameIdentifier: gameIdentifier,
		version:        version,
		buildNo:        buildNo,
		replayLengthMS: replayLengthMS,
	}

	isReforged := buildNo >= 6089

	var blocks []dataBlock
	for p.remaining() > 0 {
		blockSize, ok := p.readU16LE()
		if !ok {
			break
		}
		if isReforged {
			p.skip(2)
		}
		blockDecompSize, ok := p.readU16LE()
		if !ok {
			break
		}
		if isReforged {
			p.skip(6)
		} else {
			p.skip(4)
		}
		content, ok := p.readBytes(int(blockSize))
		if !ok {
			break
		}
		if blockDecompSize == 8192 {
			contentCopy := make([]byte, len(content))
			copy(contentCopy, content)
			blocks = append(blocks, dataBlock{
				blockSize:            blockSize,
				blockDecompressedSize: blockDecompSize,
				blockContent:         contentCopy,
			})
		}
	}

	return header, subHeader, blocks
}

func decompressBlocks(blocks []dataBlock) []byte {
	var result []byte
	for _, block := range blocks {
		if len(block.blockContent) == 0 {
			continue
		}
		r, err := zlib.NewReader(bytes.NewReader(block.blockContent))
		if err != nil {
			continue
		}
		buf, err := io.ReadAll(r)
		r.Close()
		// Accept io.ErrUnexpectedEOF: some WC3 replay blocks omit the zlib Adler-32 checksum
		if (err == nil || isUnexpectedEOF(err)) && len(buf) > 0 {
			result = append(result, buf...)
		}
	}
	return result
}

func isUnexpectedEOF(err error) bool {
	if err == io.ErrUnexpectedEOF {
		return true
	}
	return strings.Contains(err.Error(), "unexpected EOF")
}

func findSubsequence(haystack, needle []byte) int {
	if len(needle) == 0 {
		return 0
	}
	for i := 0; i <= len(haystack)-len(needle); i++ {
		if bytes.Equal(haystack[i:i+len(needle)], needle) {
			return i
		}
	}
	return -1
}

File: w3grs/src/parser/raw.rs
use super::buffer::BufParser;
use flate2::read::ZlibDecoder;
use std::io::Read;

#[derive(Debug, Clone)]
pub struct Header {
    pub compressed_size: u32,
    pub header_version: String,
    pub decompressed_size: u32,
    pub block_count: u32,
}

#[derive(Debug, Clone)]
pub struct SubHeader {
    pub game_identifier: String,
    pub version: u32,
    pub build_no: u16,
    pub replay_length_ms: u32,
}

#[derive(Debug)]
pub struct DataBlock {
    pub block_size: u16,
    pub block_decompressed_size: u16,
    pub block_content: Vec<u8>,
}

pub fn parse(input: &[u8]) -> Option<(Header, SubHeader, Vec<DataBlock>)> {
    let mut p = BufParser::new(input);

    // Find "Warcraft III recorded game\0"
    let magic = b"Warcraft III recorded game";
    let start = find_subsequence(input, magic)?;
    p.set_pos(start);
    p.read_zero_term_string(); // consume the null-terminated magic string
    p.skip(4); // unknown 4 bytes

    let compressed_size = p.read_u32_le()?;
    let header_version = p.read_string_of_length_as_hex(4);
    let decompressed_size = p.read_u32_le()?;
    let block_count = p.read_u32_le()?;

    let header = Header {
        compressed_size,
        header_version,
        decompressed_size,
        block_count,
    };

    // Subheader
    let game_identifier = p.read_string_of_length_utf8(4);
    let version = p.read_u32_le()?;
    let build_no = p.read_u16_le()?;
    p.skip(2);
    let replay_length_ms = p.read_u32_le()?;
    p.skip(4);

    let subheader = SubHeader {
        game_identifier,
        version,
        build_no,
        replay_length_ms,
    };

    let is_reforged = build_no >= 6089;

    let mut blocks = Vec::new();
    while p.remaining() > 0 {
        let block_size = match p.read_u16_le() { Some(v) => v, None => break };
        if is_reforged { p.skip(2); }
        let block_decompressed_size = match p.read_u16_le() { Some(v) => v, None => break };
        if is_reforged { p.skip(6); } else { p.skip(4); }
        let content = match p.read_bytes(block_size as usize) { Some(b) => b.to_vec(), None => break };

        if block_decompressed_size == 8192 {
            blocks.push(DataBlock {
                block_size,
                block_decompressed_size,
                block_content: content,
            });
        }
    }

    Some((header, subheader, blocks))
}

pub fn decompress_blocks(blocks: &[DataBlock]) -> Vec<u8> {
    let mut result = Vec::new();
    for block in blocks {
        if block.block_content.is_empty() {
            continue;
        }
        let mut decoder = ZlibDecoder::new(block.block_content.as_slice());
        let mut buf = Vec::new();
        if decoder.read_to_end(&mut buf).is_ok() && !buf.is_empty() {
            result.extend_from_slice(&buf);
        }
    }
    result
}

fn find_subsequence(haystack: &[u8], needle: &[u8]) -> Option<usize> {
    haystack.windows(needle.len()).position(|w| w == needle)
}
